### ISSUE TITLE
How to get the groups based on connectionId?

### ISSUE THREAD

**Issue by icistrate** [aspnet/SignalR#2395]
Hi, 
I am using RC1-final.
It seems I have troubles to find a way to get a list of groups based on connectionId. 
If there is a call `this.Groups.AddToGroupAsync(connectionId, groupName);` , why wouldn't exist such call I am looking for ? Even a `IEnumerable<string> GetGroups(connectionId)` should suffice.
Thanks

**Comment by icistrate** [comment]
Looks like I need a kind of event aggregator with SignalR , so I thought I could use the Groups to group the subscribers. So, when a client will call a method of a hub (publish an event), the group(s) where client belongs to will notify all the rest of the clients (handle the event). 
Or is there any other recommendation (maybe another library I could use for such purpose)?

**Comment by KennethInfoEasy** [comment]
You could have a connectionList which keeps as collection of the groups for each connection id,, then invoke a singalr method which sends the message to each of the groups.

**Comment by icistrate** [comment]
Yes, this is what I started with now.. at least I know now that there is nothing built-in.

**Comment by anurse** [comment]
It's also impossible to get a list of group members once you introduce a scale-out system like Redis or the Azure SignalR Service. Each server only knows about the clients connected directly to it.

Groups are **not** a stateful data structure you should use for tracking data like this, they are simply an address that you can send to. Tracking the members across all servers would be extremely complicated and beyond the scope of SignalR, so the system is designed to ensure you don't depend on that information.

If you can provide more detail about how you'd like to use groups, we can provide better guidance as to how you can solve the underlying problem. It sounds like a combination of SignalR Groups plus some shared state (in a Database, or other storage system) would work well here.

**Comment by systemidx** [comment]
@anurse I think that helps identify what the intended purpose of a group is, thanks. That also answered a question as to how it interacts with scaling at n+1 nodes (like an Azure App Service). 

To use a chat room analogy (in a multi-node architecture), would you suggest using a pub/sub with redis type of setup (or even logging connection IDs to SQL) for creating logical groupings of connected users?

At the end of the day, I'll need to group connected clients (across multiple nodes) by a logical mechanism (room ID, location, etc) and broadcast to them based on some criteria. In a clustered set-up, how would you handle broadcasting to clients connected on different nodes? Would there need to be some intercommunication set up in an AMQP exchange for the Azure SignalR services to figure out what to send and to whom?

**Comment by anurse** [comment]
So if you just need to pub/sub to logical groupings of users, that's what Groups are for. What we don't provide is a way to enumerate the connections in a group, or to enumerate the groups that a connection is a member of. This is primarily because of the way scale-out tries to avoid storing state in the backplane. Our scale-out systems don't store any state, they simply subscribe to well-known channels (based on group names) in order to get messages.

If you need to track the "groups" that a connection is a member of in order to figure out where else to broadcast a message, you can store that information in the `Context.Items` dictionary yourself when you add the connection to the group.

**Comment by systemidx** [comment]
Hmm. That seems fairly limited. I could probably get away with not needing to enumerate the connections in a group (since that's what broadcasting is for, aside from sending things to myself [which I think GroupExcept works]), but I'd still need a way to detect if there are even any clients connected to a particular group. 

For example, I'd like to be able to detect if a group (or a "chat room") has any users, aside myself, connected to it.

**Comment by anurse** [comment]
> For example, I'd like to be able to detect if a group (or a "chat room") has any users, aside myself, connected to it.

For what purpose? It's not possible to know this accurately when there is more than one server without introducing a lot of state management.

**Comment by systemidx** [comment]
To know whether or not you would be talking to an empty room. Or to "close"
the room once the last connection closes.

If there aren't any plans to add some kind of state management to SignalR,
that's fine, I can roll my own. Just don't want to wait on something that
isn't coming and just use it as a websocket server in azure.

I suppose I would like to know what the typical use cases are for the
existing group construct, as I seem to be missing the mark on that.

**Comment by davidfowl** [comment]
You still need your own state management regardless. What if you wanted to persist the current room state in a chat system across application restarts (like any decent chat application would)? That would be stored in the database. All SignalR is doing is giving you a way to do multicast efficiently for a number of connections in a group.

Now beyond groups, we'd like to figure out how to optionally add user presence tracking in some first class way that worked properly when you have more than one server but we don't have any concrete plans yet. The Azure SignalR service could do it trivially since it has a global view of all connections but we don't have a way to surface that information to the applications meaningfully as yet.

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
